### PR TITLE
Integration: Refactor event queue

### PR DESCRIPTION
**Pull request by KacperKluka**
This is an aggregate PR of the whole event queue refactoring work. All refactoring PRs should be landed on `integration/refactor-event-queue` branch. After all the work is done, this PR will be landed on the `main` branch and the refactoring work will be completed. Until that, the PR should stay in the "draft state".
Additionally, this PR will show the state of tests for the refactoring work.

### CODE REVIEW

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/ConnectionReadyWorker.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        val trackableStateFlow = properties.trackableStateFlows[trackable.id] ?: MutableStateFlow(trackableState)
+        properties.trackableStateFlows[trackable.id] = trackableStateFlow
+        corePublisher.updateTrackableStateFlows(properties)
+        properties.trackableStates[trackable.id] = trackableState
+
+        val successResult = Result.success(trackableStateFlow.asStateFlow())
+        callbackFunction(successResult)
+        properties.duplicateTrackableGuard.finishAddingTrackable(trackable, successResult)
+
+        return SyncAsyncResult()
```
Comment (ikbalkaya) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/ConnectionReadyWorker.kt:57]: I think this function is a bit long. Could it be splitted to smaller chunks?

Comment (ikbalkaya) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/ConnectionReadyWorker.kt:57]: I think this function is a bit long. Could it be splitted to smaller chunks?

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/DisconnectSuccessWorker.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        // When we remove the last trackable then we should stop location updates
+        if (properties.trackables.isEmpty() && properties.isTracking) {
+            corePublisher.stopLocationUpdates(properties)
+        }
+
+        properties.lastChannelConnectionStateChanges.remove(trackable.id)
+
+        callbackFunction(Result.success(Unit))
```
Comment (ikbalkaya) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/DisconnectSuccessWorker.kt:54]: I think this function also can be split in smaller functions.

Comment (ikbalkaya) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/DisconnectSuccessWorker.kt:54]: I think this function also can be split in smaller functions.

**Review by ikbalkaya — CHANGES_REQUESTED** [review]
I was only able to skim read and spot some areas of improvement as I was involved in this. For now I requested some changes for some functions that can be shortened in smaller chunks to increase readability.

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/DisconnectSuccessWorker.kt (hunk 2 above)
Comment (KacperKluka) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/DisconnectSuccessWorker.kt]: Good idea :+1: https://github.com/ably/ably-asset-tracking-android/pull/607/commits/ac6e25cee47556aa01fa1a04cc14c70000d25653

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/ConnectionReadyWorker.kt (hunk 1 above)
Comment (KacperKluka) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/ConnectionReadyWorker.kt]: This looks like a duplicate of https://github.com/ably/ably-asset-tracking-android/pull/607#discussion_r834080395 :wink:

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/DisconnectSuccessWorker.kt (hunk 2 above)
Comment (KacperKluka) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/DisconnectSuccessWorker.kt]: This looks like a duplicate of https://github.com/ably/ably-asset-tracking-android/pull/607#discussion_r834082234 :wink:

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/ConnectionReadyWorker.kt (hunk 1 above)
Comment (ikbalkaya) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/ConnectionReadyWorker.kt]: I realised that happening when I submitted my review, I think there is a bug with Github 🙄

**Review by ikbalkaya — APPROVED** [review]
As long as it passes integration tests looks good to me

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/results/WorkResult.kt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * This sealed class represents a base result from concrete  [Worker] instances after they finish doing their work.
+ * It's also intended to be returned by [AsyncWork] coroutines.
+ * **/
+internal sealed class WorkResult
+
+/**
+ * A special [WorkResult] that contains an optional [syncWorkResult] and [asyncWork]
+ *
+ * @param syncWorkResult : Any work result that has resulted from synchronous work. This work result is expected to be
```
Comment (QuintinWillison) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/results/WorkResult.kt:23]: Minor thing but `@param` syntax differs from the argument below. I would change to:

```suggestion
 * @param syncWorkResult any work result that has resulted from synchronous work. This work result is expected to be
```

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/EventWorkerQueue.kt [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import com.ably.tracking.publisher.workerqueue.results.SyncAsyncResult
+import com.ably.tracking.publisher.workerqueue.results.WorkResult
+import com.ably.tracking.publisher.workerqueue.workers.Worker
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
+
+/**
+ * An implementation of [WorkerQueue] that executes incoming workers defined in [execute] method. Please note
+ * that this is currently an acting bridge between the older event queue (CorePublisher) . All methods must be called
```
Comment (QuintinWillison) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/EventWorkerQueue.kt:15]: I'm not sure I understand this sentence:

> Please not that this is currently an acting bridge between the older event queue (CorePublisher).

Use of 'between' would suggest there also needs to be an 'and' - e.g. a bridge _between_ A _and_ B. 😕

**Review by QuintinWillison — APPROVED** [review]
I'm happy to merge this in its current state, despite minor observations (which could be handled in a subsequent PR in order not to block this one).

I've not reviewed line by line, so definitely a lighter weight review from me.

I notice use of the `Impl` suffix on at least one class. Generally I would prefer to avoid abbreviations if possible, plus I don't feel that - as a suffix, in this form - it's that friendly to maintainers. Could you have considered a prefix like `Default`? [I wrote [a similar comment elsewhere](https://github.com/ably/kafka-connect-ably/pull/50#discussion_r835094816) last week]

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/EventWorkerQueue.kt (hunk 4 above)
Comment (KacperKluka) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/EventWorkerQueue.kt]: This is an old description from the time when we were using both the old EventQueue and the new WorkerQueue, I've forgotten to update it, thanks for catching it :grin: I've updated the docs and renamed the class as this has no longer anything in common with `Event`s :wink: https://github.com/ably/ably-asset-tracking-android/pull/607/commits/020ec36779fab45ac9edbc312a5d148062bb85a4

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/results/WorkResult.kt (hunk 3 above)
Comment (KacperKluka) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/results/WorkResult.kt]: I've gone through the docs and aligned their style to match it project-wise https://github.com/ably/ably-asset-tracking-android/pull/607/commits/afb536853418b5bfdbdf9691ebe0b0737536adda

**Comment by KacperKluka** [comment]
@QuintinWillison thanks for the suggestion, I've replaced the `Impl` suffix with `Default` prefix in classes :wink: https://github.com/ably/ably-asset-tracking-android/pull/607/commits/aaca9be261cbd8f416b049e8e16d46b26f209c00

**Review by QuintinWillison — APPROVED** [review]

**Pull request merged**